storage/pebble/qcs.go

```diff
@@ -1,33 +1,36 @@
-package badger
+package pebble
 
 import (
-	"github.com/dgraph-io/badger/v2"
+	"sync"
+
+	"github.com/cockroachdb/pebble"
 
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/module"
 	"github.com/onflow/flow-go/module/metrics"
 	"github.com/onflow/flow-go/storage"
-	"github.com/onflow/flow-go/storage/badger/operation"
 	"github.com/onflow/flow-go/storage/badger/transaction"
+	"github.com/onflow/flow-go/storage/pebble/operation"
 )
 
 // QuorumCertificates implements persistent storage for quorum certificates.
 type QuorumCertificates struct {
-	db    *badger.DB
-	cache *Cache[flow.Identifier, *flow.QuorumCertificate]
+	storing sync.Mutex
+	db      *pebble.DB
+	cache   *Cache[flow.Identifier, *flow.QuorumCertificate]
 }
 
 var _ storage.QuorumCertificates = (*QuorumCertificates)(nil)
 
 // NewQuorumCertificates Creates QuorumCertificates instance which is a database of quorum certificates
 // which supports storing, caching and retrieving by block ID.
-func NewQuorumCertificates(collector module.CacheMetrics, db *badger.DB, cacheSize uint) *QuorumCertificates {
-	store := func(_ flow.Identifier, qc *flow.QuorumCertificate) func(*transaction.Tx) error {
-		return transaction.WithTx(operation.InsertQuorumCertificate(qc))
+func NewQuorumCertificates(collector module.CacheMetrics, db *pebble.DB, cacheSize uint) *QuorumCertificates {
+	store := func(_ flow.Identifier, qc *flow.QuorumCertificate) func(storage.PebbleReaderBatchWriter) error {
+		return storage.OnlyWriter(operation.InsertQuorumCertificate(qc))
```

Comment:
## ⚠️ 

Previously, badger guaranteed that we would only ever write one QC for a block into the database. In other words, a QC for a block that was stored in the database would never change. 

This now changes:
If `operation.InsertQuorumCertificate` is ever called _outside_ of the `QuorumCertificates` storage abstraction layer, we are risking to update (overwrite) the QC for a block. The reason is that you included a mutex here, which only protects dirty writes mediated by `QuorumCertificates`, but the lower-level `operation.InsertQuorumCertificate` remains easily accessible for any logic.

I think the way we are using QCs, overwriting the QC would be fine. We only require _a_ valid QC, but my gut feeling is that it doesn't need to be consistently the same. Nevertheless, this requires careful confirmation and proper documentation.